# Problem 1: PLU and determinants

In [ ]:
A = matrix(QQ, [[2 , 1 , 1 , 1 ],
                [-2 , -1 , -3 , 0 ],
                [0 , 1 , 2 , 2 ],
                [4 , 1 , -2 , -2]])

In [ ]:
E1 = identity_matrix(QQ,4); E1[0:,0]=-A[0:,0]/A[0,0];E1[0,0]=1; A1=E1*A
show(E1,A1)

E2 = identity_matrix(QQ,4); E2[1,1]=0;E2[1,2]=1;E2[2,1]=1;E2[2,2]=0; A2=E2*A1
show(E2,A2)

E3 = identity_matrix(QQ,4); E3[1:,1]=-A2[1:,1]/A2[1,1];E3[1,1]=1; A3=E3*A2
show(E3,A3)

E4 = identity_matrix(QQ,4); E4[2:,2]=-A3[2:,2]/A3[2,2];E4[2,2]=1; A4=E4*A3
show(E4,A4)

[ 1  0  0  0]
[ 1  1  0  0]
[ 0  0  1  0]
[-2  0  0  1] [ 2  1  1  1]
[ 0  0 -2  1]
[ 0  1  2  2]
[ 0 -1 -4 -4]

[1 0 0 0]
[0 0 1 0]
[0 1 0 0]
[0 0 0 1] [ 2  1  1  1]
[ 0  1  2  2]
[ 0  0 -2  1]
[ 0 -1 -4 -4]

[1 0 0 0]
[0 1 0 0]
[0 0 1 0]
[0 1 0 1] [ 2  1  1  1]
[ 0  1  2  2]
[ 0  0 -2  1]
[ 0  0 -2 -2]

[ 1  0  0  0]
[ 0  1  0  0]
[ 0  0  1  0]
[ 0  0 -1  1] [ 2  1  1  1]
[ 0  1  2  2]
[ 0  0 -2  1]
[ 0  0  0 -3]

In [ ]:
# 1a)
P=E2; L= P*E1.inverse()*P*E3.inverse()*E4.inverse();U=A4
show(P*L*U, ' = ', P, L,U)

[ 2  1  1  1]
[-2 -1 -3  0]
[ 0  1  2  2]
[ 4  1 -2 -2] ' = ' [1 0 0 0]
[0 0 1 0]
[0 1 0 0]
[0 0 0 1] [ 1  0  0  0]
[ 0  1  0  0]
[-1  0  1  0]
[ 2 -1  1  1] [ 2  1  1  1]
[ 0  1  2  2]
[ 0  0 -2  1]
[ 0  0  0 -3]

In [ ]:
# 1b
b = vector(QQ,[1,9,0,0])
w = P.inverse()*b
y = L.inverse()*w
x = U.augment(y).rref()[:,4].transpose()

print 'b =',b
print 'w =',w
print 'y =',y
print 'x =', vector(x)

b = (1, 9, 0, 0)
w = (1, 0, 9, 0)
y = (1, 0, 10, -12)
x = (1, -2, -3, 4)


In [ ]:
#1c
detA = -product(U.diagonal())
print detA

-24 -12


In [ ]:
# 1d)
print '-3^4 * 12^2 = - 2^4 3^6 = ', -3^4 * detA^2

-3^4 * 12^2 = - 2^4 3^6 =  -46656


1e)<br>
$A$ orthogonal $\Leftrightarrow  A^t A = I$;   Note A need not be square<br>
$det(A^t A) = det(I)  \Leftrightarrow  det(A)^2 = 1 \Leftrightarrow  det(A) = 1 \text{ or } -1$<br>
an orthogonal matrix cannot be singular since its determinant is not zero

# Problem 2: QR

In [ ]:
v1 = matrix(QQ,[[0, 1, 2, 2]]).transpose()
v2 = matrix(QQ,[[2, -3, -2,-1]]).transpose()
v3 = matrix(QQ,[[1,1,3,3]]).transpose()

A= v1.augment(v2).augment(v3)
show(A)

[ 0  2  1]
[ 1 -3  1]
[ 2 -2  3]
[ 2 -1  3]

In [ ]:
def prj(v,w):
    return (v.transpose()*w)[0,0]/(w.transpose()*w)[0,0]*w

w1 = v1
w2 = v2 -prj(v2,w1)
w3 = 9*(v3 - prj(v3,w1) -prj(v3,w2))

#  ------------------------------------------------
# print out the results
x=var('x')
x4=matrix([[x,x,x,x],[x,x,x,x],[x,x,x,x],[x,x,x,x]])
x3=matrix([[x,x,x],[x,x,x],[x,x,x]])

W = w1.augment(w2).augment(w3)
show(x3, x4, A,W)
show (x3, W.transpose(),W.transpose()*A, W.transpose()*W)
S = diagonal_matrix( [1/3, 1/3, sqrt(2)/6 ] )
Qt = S* W.transpose()
Q = Qt.transpose()
R = Qt * A
show(S, Qt, R)

[x x x]
[x x x]
[x x x] [x x x x]
[x x x x]
[x x x x]
[x x x x] [ 0  2  1]
[ 1 -3  1]
[ 2 -2  3]
[ 2 -1  3] [ 0  2  3]
[ 1 -2  2]
[ 2  0  1]
[ 2  1 -2]

[x x x]
[x x x]
[x x x] [ 0  1  2  2]
[ 2 -2  0  1]
[ 3  2  1 -2] [ 9 -9 13]
[ 0  9  3]
[ 0  0  2] [ 9  0  0]
[ 0  9  0]
[ 0  0 18]

[        1/3           0           0]
[          0         1/3           0]
[          0           0 1/6*sqrt(2)] [           0          1/3          2/3          2/3]
[         2/3         -2/3            0          1/3]
[ 1/2*sqrt(2)  1/3*sqrt(2)  1/6*sqrt(2) -1/3*sqrt(2)] [          3          -3        13/3]
[          0           3           1]
[          0           0 1/3*sqrt(2)]

In [ ]:
show(A, ' = ', Q, R)

[ 0  2  1]
[ 1 -3  1]
[ 2 -2  3]
[ 2 -1  3] ' = ' [           0          2/3  1/2*sqrt(2)]
[         1/3         -2/3  1/3*sqrt(2)]
[         2/3            0  1/6*sqrt(2)]
[         2/3          1/3 -1/3*sqrt(2)] [          3          -3        13/3]
[          0           3           1]
[          0           0 1/3*sqrt(2)]

In [ ]:
# 2b)
print 'least squares solution R x = Q.transpose b'
b=matrix(QQ,4,1,[5, 5, 1, 3])
x_star = R.inverse()*Qt*b
show(b, '-> x =', x_star)

b_c = A * x_star
b_n = b - b_c
show( 'bc = ', b_c,' bn =', b_n)

least squares solution R x = Q.transpose b


[5]
[5]
[1]
[3] '-> x =' [-16]
[ -3]
[ 10]

'bc = ' [4]
[3]
[4]
[1] ' bn =' [ 1]
[ 2]
[-3]
[ 2]

4c)<br>
x_star is the solution of $argmin_x ||A x - b||$<br>

show the decomposition of b into $b_c+b_n$ (codomain of A turned sideways)

# Problem 3: SVD

In [ ]:
A = matrix(QQ,[[-3,4],[-2,11],[-10,5],[0,0]])

AtA = A.transpose()*A
show('A=',A, ',  At A =', AtA, ';   50 is an eigenvalue: ', (AtA - 50*identity_matrix(QQ,2)).rref())
print 'Charpoly ', factor(AtA.charpoly())

'A=' [ -3   4]
[ -2  11]
[-10   5]
[  0   0] ',  At A =' [113 -84]
[-84 162] ';   50 is an eigenvalue: ' [   1 -4/3]
[   0    0]

Charpoly  (x - 225) * (x - 50)


Since $\lambda_1 = 50$, $\lambda_2 = tr(A)-50 = 113+162-50 = 225$

To get $V$, we need the nullspace and column space of $A^t A - \lambda I$ (the special case of a 2x2 matrix):
$$
A^t A - 50 I = \begin{pmatrix} 63 & -84 \\ -84 & 112 \end{pmatrix}
$$
Thus eigenvector for 50 is (84 63) -> ( 4, 3)/5<br>
Eigenvector for 225 must be orthogonal to this (or first col of $(A^t A - 50 I)$ since the matrix is 2x2) -> (3,-4)/5

In [ ]:
V = AtA.right_eigenmatrix()[1]*matrix(QQ, [[3,0],[0,4]])/5
show('V =', 1/5,5*V)

'V =' 1/5 [ 3  4]
[-4  3]

In [ ]:
Sigma_r = diagonal_matrix([sqrt(225),sqrt(50)])
U_r     = A*V*Sigma_r.inverse()
show(A, ' = ', 1/6,6*U_r,5,1/5*Sigma_r, 1/5,5*V.transpose())

[ -3   4]
[ -2  11]
[-10   5]
[  0   0] ' = ' 1/6 [        -2          0]
[        -4  3*sqrt(2)]
[        -4 -3*sqrt(2)]
[         0          0] 5 [      3       0]
[      0 sqrt(2)] 1/5 [ 3 -4]
[ 4  3]

In [ ]:
print 'PseudoInverse'
A_pseudo = V*Sigma_r.inverse()*U_r.transpose()
show( 1/5,5*V,1/30,30*Sigma_r.inverse(), 1/6,6*U_r.transpose(),' = ',1/(5*30*3),5*30*3*A_pseudo)

PseudoInverse


1/5 [ 3  4]
[-4  3] 1/30 [        2         0]
[        0 3*sqrt(2)] 1/6 [        -2         -4         -4          0]
[         0  3*sqrt(2) -3*sqrt(2)          0] ' = ' 1/450 [ -6  24 -48   0]
[  8  43 -11   0]

In [ ]:
U=(U_r*matrix([[45,0],[0,10]])).augment(identity_matrix(QQ,4))
E1 = matrix(QQ,identity_matrix(QQ,4)); E1[1,0]=-2;E1[2,0]=-2
show(U)
U1=E1*U
show(E1,U1)
E2 = matrix(QQ, identity_matrix(QQ,4)); E2[2,1]=1
U2=E2*U1
show( E2,U2)

[       -15          0          1          0          0          0]
[       -30  5*sqrt(2)          0          1          0          0]
[       -30 -5*sqrt(2)          0          0          1          0]
[         0          0          0          0          0          1]

[ 1  0  0  0]
[-2  1  0  0]
[-2  0  1  0]
[ 0  0  0  1] [       -15          0          1          0          0          0]
[         0  5*sqrt(2)         -2          1          0          0]
[         0 -5*sqrt(2)         -2          0          1          0]
[         0          0          0          0          0          1]

[1 0 0 0]
[0 1 0 0]
[0 1 1 0]
[0 0 0 1] [      -15         0         1         0         0         0]
[        0 5*sqrt(2)        -2         1         0         0]
[        0         0        -4         1         1         0]
[        0         0         0         0         0         1]

In [ ]:
U = (U_r*matrix([[45,0],[0,10]])).augment(U2[2,2:].transpose())
U = U.augment(vector([0,0,0,1]))
U = U * diagonal_matrix([2/90,1/sqrt(100), 1/sqrt(18),1])
#U.transpose()*U
show( 'U=',1/6,6* U, ',  Sigma=', Sigma_r.augment(matrix([[0,0],[0,0]]) ).transpose(),  ',   V=',1/5,5*V)

'U=' 1/6 [        -2          0 -4*sqrt(2)          0]
[        -4  3*sqrt(2)    sqrt(2)          0]
[        -4 -3*sqrt(2)    sqrt(2)          0]
[         0          0          0          6] ',  Sigma=' [       15         0]
[        0 5*sqrt(2)]
[        0         0]
[        0         0] ',   V=' 1/5 [ 3  4]
[-4  3]

In [ ]:
b = matrix(4,1,[0,-2,1,0])
x_star = A_pseudo*b
b_parallel = A*x_star
b_perp     = b - b_parallel

show('xstar     = ',1/450,450*x_star)
show('bparallel = ', 1/18,18*b_parallel)
show('bperp     = ', 1/18,18*b_perp)

'xstar     = ' 1/450 [-96]
[-97]

'bparallel = ' 1/18 [ -4]
[-35]
[ 19]
[  0]

'bperp     = ' 1/18 [ 4]
[-1]
[-1]
[ 0]

3c) rank(A) = 2<br>
3d) Condition Number = $\sqrt\frac{225}{50}$<br>
3e)<br>
cols of V are a basis for $\mathscr{R}(A)$<br>
cols of $U_r$ are a basis for $\mathscr{C}(A)$<br>
remaining cols of $U$ are a basis for $\mathscr{N}(A^t)$<br>
basis for $\mathscr{N}(A)$ is empty

# Problem 4: Extrema and Level Curves

In [ ]:
A = matrix(QQ, [[9,0,2],[0,5,0],[2,0,6]])
x,y,z=var('x,y,z')
v=matrix(3,1,[x,y,z])
show(A)
expand(v.transpose()*A*v)

[9 0 2]
[0 5 0]
[2 0 6]

[9*x^2 + 5*y^2 + 4*x*z + 6*z^2]

In [ ]:
factor(A.charpoly())

(x - 10) * (x - 5)^2

In [ ]:
Lambda,S=A.right_eigenmatrix(); S=S*diagonal_matrix([2/sqrt(5),1/sqrt(5),1])
show('A =', 1/sqrt(5),5/(sqrt(5))*S, ',  Lambda =', 5, 1/5*Lambda, ',  St =', 1/sqrt(5),5/(sqrt(5))*S.transpose()())
print('Note S.inverse=S.transpose')

'A =' 1/5*sqrt(5) [      2       1       0]
[      0       0 sqrt(5)]
[      1      -2       0] ',  Lambda =' 5 [2 0 0]
[0 1 0]
[0 0 1] ',  St =' 1/5*sqrt(5) [      2       0       1]
[      1       0      -2]
[      0 sqrt(5)       0]

Note S.inverse=S.transpose


4a) $f(x_1,x_2,x_3) = 10 \tilde{x_1}^2 + 5 \tilde{x_2}^2 + 5 \tilde{x_3}^2$ using the substitution $x = S \tilde{x}$<br>
$\quad f(x_1,x_2,x_3) = -1$ therefore has no solution since $f(x_1,x_2,x_3) \ge 0$<br>
4b) The matrix A is positive definite<br>
4c) The critical point is a minimum since A is positive definite<br>
Second test: the pivots of A are 9,5, and 6-4/9 which are all positive

# Problem 5: Iterative Solutions

In [ ]:
A=matrix(QQ,3,3,[5,-1,0,-1,2,0,0,0,1])
b=matrix(QQ,3,1,[4,1,1])
show(A,b)

[ 5 -1  0]
[-1  2  0]
[ 0  0  1] [4]
[1]
[1]

In [ ]:
print '5a)  Jacobi'
S=matrix(QQ,3,3,[5,0,0, 0,2,0, 0,0,1])
T=S-A
show('S=',S,' T=', T, ';  Sinv T=', S.inverse()*T, ', Sinv b=',S.inverse()*b)
print( factor((S.inverse()*T).charpoly())), ' spectral radius sqrt(10)/10'

print '5b) Gauss-Seidel'
S=matrix(QQ,3,3,[5,-1,0,0,2,0,0,0,1])
T=S-A
show('S=',S,' T=', T, ';  Sinv T=', S.inverse()*T, ', Sinv b=',S.inverse()*b)
print( factor((S.inverse()*T).charpoly())), ' spectral radius 1/10'

5a)  Jacobi


'S=' [5 0 0]
[0 2 0]
[0 0 1] ' T=' [0 1 0]
[1 0 0]
[0 0 0] ';  Sinv T=' [  0 1/5   0]
[1/2   0   0]
[  0   0   0] ', Sinv b=' [4/5]
[1/2]
[  1]

x * (x^2 - 1/10)  spectral radius sqrt(10)/10
5b) Gauss-Seidel


'S=' [ 5 -1  0]
[ 0  2  0]
[ 0  0  1] ' T=' [0 0 0]
[1 0 0]
[0 0 0] ';  Sinv T=' [1/10    0    0]
[ 1/2    0    0]
[   0    0    0] ', Sinv b=' [9/10]
[ 1/2]
[   1]

(x - 1/10) * x^2  spectral radius 1/10


5c) Since $\frac{1}{10} \lt \frac{\sqrt{10}}{10}$ Gauss-Seidel converges faster

In [ ]:
print '5d)'
show(S.inverse()*T*matrix(QQ,3,1,[0,1,0])+S.inverse()*b)

5d)


[9/10]
[ 1/2]
[   1]

# Problem 6  Linear Transformations

In [ ]:
v1=matrix(QQ,3,1,[1,1,0])
v2=matrix(QQ,3,1,[-2,1,1])
show(v1,v2)
print 'Orthogonal vector'
v3 = ( v1.augment(v2).augment(identity_matrix(QQ,3)).rref()[2,2:]).transpose()
show(v3)

[1]
[1]
[0] [-2]
[ 1]
[ 1]

Orthogonal vector


[ 1]
[-1]
[ 3]

In [ ]:
print '6b'
print 'Orthogonal projection onto span(v3)'
P2= v3*(v3.transpose()*v3).inverse()*v3.transpose()
show(1/11, 11*P2)
print 'orthogonal projection onto v1,v2 plane'
P1=identity_matrix(QQ,3)-P2
show(1/11,11*P1)

6b
Orthogonal projection onto span(v3)


1/11 [ 1 -1  3]
[-1  1 -3]
[ 3 -3  9]

orthogonal projection onto v1,v2 plane


1/11 [10  1 -3]
[ 1 10  3]
[-3  3  2]

In [ ]:
print '6c)'
c3 = cos(pi/6); s3=sin(pi/6)
R = matrix( 3,3, [c3,-s3,0, s3,c3,0,  0,0,1])
show(1/2,2*R)

6c)


1/2 [sqrt(3)      -1       0]
[      1 sqrt(3)       0]
[      0       0       2]

In [ ]:
print '6d)'
M = R*P1
show(1/22,22*M)

6d)


1/22 [10*sqrt(3) - 1   sqrt(3) - 10 -3*sqrt(3) - 3]
[  sqrt(3) + 10 10*sqrt(3) + 1  3*sqrt(3) - 3]
[            -6              6              4]

6e)
Since $P_1$ is not 1-1 and not onto, neither is $R P_1$:<br>
$y = P_1 x_1, y = P_1 x_2, x_1 \ne x_2  \Rightarrow R y = R P_1 x_1 = R P_2 x_2, x_1 \ne x_2$<br>
$b \in span\{v_3\} \Rightarrow y = P_2 b$ has no solution. Since $R^{-1}$ exists, $ R b = R P_1 x \Rightarrow b = P_1 x$, so this equation has no solution.

# Problem 7: Difference Equation
$u_{n+3}= 2 u_{n+2} + u_{n+1}-2 u_n, \quad u_2 = 1, u_1 = 0, u_0 = 1$

In [ ]:
print 'add the equalities u_{n+2}=u{n+2}, u{n+1}=u_{n+1}  to get'
u_n3,u_n2,u_n1,u_n=var('u_n3,u_n2,u_n1,u_n')

A = matrix(QQ, 3,3, [2,1,-2, 1,0,0, 0,1,0])
show( matrix(3,1,[u_n3,u_n2,u_n1]), '=', A, matrix(3,1,[u_n2,u_n1, u_n]))
print 'Note the eigenvector (1,1,1) for lambda_1=1, and thus lambda_2+lambda_3=1, lambda_2*lambda_3=-2'

add the equalities u_{n+2}=u{n+2}, u{n+1}=u_{n+1}  to get


[u_n3]
[u_n2]
[u_n1] '=' [ 2  1 -2]
[ 1  0  0]
[ 0  1  0] [u_n2]
[u_n1]
[ u_n]

Note the eigenvector (1,1,1) for lambda_1=1, and thus lambda_2+lambda_3=1, lambda_2*lambda_3=-2


In [ ]:
Lambda,S=A.eigenmatrix_right()
show( A, '=', 1/4, 4*S, Lambda, 1/6,6*S.inverse())

[ 2  1 -2]
[ 1  0  0]
[ 0  1  0] '=' 1/4 [ 4  4  4]
[ 2  4 -4]
[ 1  4  4] [ 2  0  0]
[ 0  1  0]
[ 0  0 -1] 1/6 [ 8  0 -8]
[-3  3  6]
[ 1 -3  2]

In [ ]:
n=var('n')
show( 'un3 =', Lambda.diagonal(),'^n  ',1/6, 6*S.inverse(), matrix(3,1,[1,0,1]),'  -> un3 = ', (-1)^(n+1))
print 'which DNE as n -> infinity'

'un3 =' [2, 1, -1] '^n  ' 1/6 [ 8  0 -8]
[-3  3  6]
[ 1 -3  2] [1]
[0]
[1] '  -> un3 = ' (-1)^(n + 1)

which DNE as n -> infinity


# Problem 8: ODE

$y'=A y, A=S \Lambda S^{-1} \Leftrightarrow S^{-1} y'= \Lambda S^{-1} y$<br>
Let $\tilde{y} = S^{-1} y$ then $\tilde{y}' = \Lambda \tilde{y} \Leftrightarrow \tilde{y} = e^{\Lambda t} \tilde{y_0}$

Substituting $y = S \tilde{y}$, we get  $y = S e^{\Lambda t} S^{-1} y_0$

In [ ]:
S    = matrix(QQ, 3,3, [21,-10,4, -10,10,-3, 4,-3,1])
e_lt = diagonal_matrix( [e^(x/10),e^(-x/10), e^(-8/10*x)])
y0   = matrix(3,1,[1,0,0])

show( 'e^(Lambda t) = ', e_lt )
print 'Hence, the solution is y=S e^(Lambda t) S.inverse y0'

show( 'y =', S, e_lt, S.inverse(), y0)

print 'This works out to'
show('y =',  S * e_lt * S.inverse() * matrix(3,1,[1,0,0]))

'e^(Lambda t) = ' [ e^(1/10*x)           0           0]
[          0 e^(-1/10*x)           0]
[          0           0  e^(-4/5*x)]

Hence, the solution is y=S e^(Lambda t) S.inverse y0


'y =' [ 21 -10   4]
[-10  10  -3]
[  4  -3   1] [ e^(1/10*x)           0           0]
[          0 e^(-1/10*x)           0]
[          0           0  e^(-4/5*x)] [  1  -2 -10]
[ -2   5  23]
[-10  23 110] [1]
[0]
[0]

This works out to


'y =' [ 21*e^(1/10*x) + 20*e^(-1/10*x) - 40*e^(-4/5*x)]
[-10*e^(1/10*x) - 20*e^(-1/10*x) + 30*e^(-4/5*x)]
[   4*e^(1/10*x) + 6*e^(-1/10*x) - 10*e^(-4/5*x)]

The limit as $x \rightarrow \infty$ does not exist since $e^{x/10} \rightarrow \infty$, 

In [ ]:
show( 'sin( Ax ) =', S, diagonal_matrix( [sin(x/10),sin(-x/10), sin(-8/10*x)]), S.inverse())

'sin( Ax ) =' [ 21 -10   4]
[-10  10  -3]
[  4  -3   1] [ sin(1/10*x)            0            0]
[           0 sin(-1/10*x)            0]
[           0            0  sin(-4/5*x)] [  1  -2 -10]
[ -2   5  23]
[-10  23 110]

# Problem 9
As set of vectors $B = \{v_1, v_2, \dots v_n\}$ is linearly independent iff
$\sum_{i=1}^n { \alpha_i v_i } = 0 \Rightarrow \alpha_i = 0$ for $i=1,2, \dots n$.

The set of vectors $B$ forms a basis for a vector space $V$ iff a) the vectors are linearly independent
and b) span $V$.

Given a matrix $A$ of size $M \times N$, the SVD $A = U \Sigma V^t$, where
* $\Sigma = \begin{pmatrix} \Sigma_r & 0 \\ 0 & 0 \end{pmatrix}$ of size $M \times N$,
where $\Sigma_r$ is a diagonal matrix of size $r \times r$ with non-zero entries in decreasing order on the diagonal. $r = rank(A)$.
* $U$ and $V$ are orthogonal matrices of size $M \times M$ and $N \times N$ respectively.
  * The first $r$ columns of $V$ are a basis for $\mathscr{R}(A)$, the remaining columns
    are  a basis for $\mathscr{N}(A)$
  * the first $r$ columns of $U$ are a basis for $\mathscr{C}(A)$, the remaining columns
    are  a basis for $\mathscr{N}(A^t)$.

# Problem 10
We saw three solution methods for the projection problem $argmin_x || A x - b ||$:
* using the normal equation: solve $A^t A x = A^t b$
* using the QR decomposition: $A = Q R$; solve $R x = Q^t b$
* using the pseudoinverse $A^\dagger$: compute $x = A^\dagger b$.

The vector $b$ decomposes into $b = b_c + b_n$, where $b_c \in \mathscr{C}(A)$ and $b_n \in \mathscr{N}(A^t)$:<br>
$\quad b_c = A x$<br>
$\quad b_n = b - b_c$.